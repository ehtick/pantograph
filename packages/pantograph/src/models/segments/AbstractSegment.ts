import type { Vector } from "../../definitions";
import { BoundingBox } from "../BoundingBox";
import type { TransformationMatrix } from "../TransformationMatrix";
import { reprVector } from "../../vectorOperations";
import { Transformable } from "../utils/Transformable";

export abstract class AbstractSegment<
  T extends AbstractSegment<T>,
> extends Transformable<T> {
  constructor(
    public firstPoint: Vector,
    public lastPoint: Vector,
  ) {
    super();
    this.firstPoint = firstPoint;
    this.lastPoint = lastPoint;
  }

  readonly precision: number = 1e-9;

  abstract segmentType: string;

  get repr() {
    return `${this.segmentType} ${reprVector(this.firstPoint)} - ${reprVector(
      this.lastPoint,
    )}`;
  }
  get info() {
    return this.repr;
  }

  abstract get midPoint(): Vector;

  abstract get boundingBox(): BoundingBox;

  abstract clone(): T;

  abstract reverse(): T;

  abstract isSame(other: AbstractSegment<any>): boolean;

  abstract distanceFrom(element: Vector): number;

  abstract isOnSegment(point: Vector): boolean;

  abstract tangentAt(point: Vector): Vector;

  abstract get tangentAtFirstPoint(): Vector;
  abstract get tangentAtLastPoint(): Vector;

  abstract splitAt(points: Vector[] | number[]): T[];

  abstract transform(matrix: TransformationMatrix): T;
}
