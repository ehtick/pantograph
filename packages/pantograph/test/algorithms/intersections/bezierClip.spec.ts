import { describe, it, expect } from "vitest";

import {
  bezierClip,
  /*
    * used for debugging
  fatLineFromCurve,
  fatLineIntersections,
  perpendicularFatLineFromCurve,
  perpendicularFatLineFromCurve,
  */
} from "../../../src/algorithms/intersections/bezierClip";
import { Line } from "../../../src/models/segments/Line";
import { CubicBezier } from "../../../src/models/segments/CubicBezier";
import {
  add,
  perpendicular,
  scalarMultiply,
} from "../../../src/vectorOperations";
import { QuadraticBezier } from "../../../src/models/segments/QuadraticBezier";

//import { debugImg, dpnt } from "../../debug";

function showFatLine(fatLine: any) {
  const line = new Line(fatLine.firstPoint, fatLine.lastPoint);
  const perp = perpendicular(line.tangentAtFirstPoint);

  const p1 = scalarMultiply(perp, fatLine.negativeThickness);
  const p2 = scalarMultiply(perp, fatLine.positiveThickness);

  return [
    { shape: line, color: "grey" },
    {
      shape: new Line(add(line.firstPoint, p1), add(line.lastPoint, p1)),
      color: "red",
    },
    {
      shape: new Line(add(line.firstPoint, p2), add(line.lastPoint, p2)),
      color: "blue",
    },
  ];
}

describe("bezierClip", () => {
  describe("cubic bezier", () => {
    it("should find the intersection of two bezier curve", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1, 0], [0, 1]);

      const rotationPoint = curve1.paramPoint(0.6);

      const curve2 = curve1.rotate(70, rotationPoint);
      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(1);
      expect(intersections[0]).toBeVector(rotationPoint);
    });

    it("should find no intersection when they do not intersect", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1, 0], [0, 1]);
      const curve2 = curve1.translateX(-1);

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(0);
    });

    it("should find two intersections of two bezier curve", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1, 0], [0, 1])
        .scale(1.5)
        .rotate(-40)
        .translate(-0.5, 0.8);
      const curve2 = new CubicBezier([-0.5, 0], [1, 0], [-1, 1], [2, 3]);

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(2);
      expect(intersections[0]).toMatchSnapshot();
    });

    it("should find three intersections of two bezier curve", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1.2, 0], [0, 2]);
      const curve2 = curve1
        .mirror("x")
        .translate(-0.1, 1.5)
        .rotate(-40)
        .translateX(-0.5);

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(3);
      expect(intersections[0]).toMatchSnapshot();
    });
  });

  describe("quadratic bezier", () => {
    it("should find the intersection of two bezier curve", () => {
      const curve1 = new QuadraticBezier([0, 0], [1, 1], [1, 0]);

      const rotationPoint = curve1.paramPoint(0.6);

      const curve2 = curve1.rotate(70, rotationPoint);
      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(1);
      expect(intersections[0]).toBeVector(rotationPoint);
    });

    it("should find no intersection when they do not intersect", () => {
      const curve1 = new QuadraticBezier([0, 0], [1, 1], [1, 0]);
      const curve2 = curve1.translateX(-1);

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(0);
    });

    it("should find two intersections of two bezier curve", () => {
      const curve1 = new QuadraticBezier([0, 0], [1, 1], [1, 0])
        .scale(1.5)
        .rotate(-40)
        .translate(-0.5, 0.6);
      const curve2 = new QuadraticBezier([-0.5, 0], [1, 0], [-1, 1]);

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(2);
      expect(intersections[0]).toMatchSnapshot();
    });
  });

  describe("cubic and quadratic bezier", () => {
    it("should find the intersection of two bezier curve", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1, 0], [0, 1]);
      const curve2 = new QuadraticBezier([0, 0], [1, 1], [1, 0]).rotate(
        90,
        [0.2, 0.3],
      );

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(1);
      expect(intersections[0]).toMatchSnapshot();
    });

    it("should find no intersection when they do not intersect", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1, 0], [0, 1]);
      const curve2 = new QuadraticBezier([-1, -1], [0, 1], [-1, 0]);

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(0);
    });

    it("should find two intersections of two bezier curve", () => {
      const curve1 = new CubicBezier([0, 0], [1, 1], [1, 0], [0, 1])
        .scale(1.5)
        .rotate(-40)
        .translate(-0.5, 0.8);
      const curve2 = new QuadraticBezier([-0.5, 0], [1, 0], [-1, 1]).translate(
        0.2,
        0.5,
      );

      const intersections = bezierClip(curve1, curve2);

      //debugImg([curve1, curve2, ...intersections.map((i) => dpnt(i))]);

      expect(intersections.length).toBe(2);
      expect(intersections).toMatchSnapshot();
    });
  });
});
