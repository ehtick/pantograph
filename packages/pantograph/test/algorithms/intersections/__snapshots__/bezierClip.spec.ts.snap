// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bezierClip > cubic and quadratic bezier > should find the intersection of two bezier curve 1`] = `
[
  0.4892354904149274,
  0.2967398029300029,
]
`;

exports[`bezierClip > cubic and quadratic bezier > should find two intersections of two bezier curve 1`] = `
[
  [
    -0.3835135358687378,
    0.709362363650092,
  ],
  [
    0.4964645457985965,
    0.821412790119727,
  ],
]
`;

exports[`bezierClip > cubic bezier > should find three intersections of two bezier curve 1`] = `
[
  0.5658880883788215,
  1.0038974747677303,
]
`;

exports[`bezierClip > cubic bezier > should find two intersections of two bezier curve 1`] = `
[
  -0.4093170772768268,
  0.7281585614889889,
]
`;

exports[`bezierClip > quadratic bezier > should find two intersections of two bezier curve 1`] = `
[
  -0.3737131235943163,
  0.49999853214942713,
]
`;
